File: calculadora.java
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.tree.*;

public class CalcApp {
    public static void main(String[] args) throws Exception {
        // Fuente de entrada: consola (stdin)
        CharStream input = CharStreams.fromStream(System.in);

        // Etapa léxica
        labeldExprLexer lexer = new labeldExprLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);

        // Etapa de análisis sintáctico
        labeldExprParser parser = new labeldExprParser(tokens);
        ParseTree tree = parser.prog();

        // Visitador para evaluar las expresiones
        EvalVisitor visitor = new EvalVisitor();
        visitor.visit(tree);
    }
}

File: calculadora.py
import sys
from antlr4 import *
from labeldExprLexer import labeldExprLexer
from labeldExprParser import labeldExprParser
from EvalVisitor import EvalVisitor

def main(argv):
    # Determinar la fuente de entrada: archivo o stdin
    if len(argv) > 1:
        data = FileStream(argv[1])
    else:
        data = InputStream(sys.stdin.read())

    # Fase léxica
    lexer = labeldExprLexer(data)
    tokens = CommonTokenStream(lexer)

    # Fase de parsing
    parser = labeldExprParser(tokens)
    tree = parser.prog()

    # Evaluador con visitor
    evaluator = EvalVisitor()
    evaluator.visit(tree)

if __name__ == "__main__":
    main(sys.argv)
